doc/acl/policy_based_metering/ACL-Policy-Based-Metering.md

```diff
@@ -0,0 +1,402 @@
+
+Policer in networking are responsible for **metering** (Monitoring the rate of traffic) and **marking** (Flagging traffic that exceeds defined limits) traffic based on predefined criteria.
+By applying policers to ACL rules, SONiC can effectively control the flow of network traffic, ensuring fairness, optimizing bandwidth utilization, and preventing network congestion.
+
+Usage examples:
+- **Security**: Policy-Based Metering can be used to guard against **network storms or DDoS attacks** by limiting traffic rates.
+- **Quality of Service (QoS)**: It helps in prioritizing critical traffic (like VoIP or video streams) while limiting non-critical traffic (for example: backup services traffic).
+- **Data Center**: ensure fair bandwidth distribution.
+
+---
+### Requirements
+#### Functional Requirements
+- Backward compatibility for existing ACL features - If policer is not set, the system will function as it did previously.
+- Ability to config policers with ACL entries.
+- Support all existing Policer types (Policer mode, meter_type).
+- Support all existing ACL types (ACL table types, ACL stages).
+### Scalability Requirements:
+- Support for multiple ACL entries with associated policers.
+- Query and validate ASIC capabilities dynamically.
+#### CLI Requirements
+- Bind policers with ACL rules.
+- Unbind policer from ACL rules.
+- Show ACLs with policers
+---
+### Architecture Design
+
+No SONiC architecture changes are required as the existing infrastructure is being used.
+
+---
+### High-Level Design
+
+#### Modules and Sub-Modules
+
+- **SWSS**
+  - ACL-Orch
+    - Add/Remove/Manage ACLs.
+    - Set or disable policer action.
+    - Query from SAI the ACL actions capability.
+    - Allow policer action only when capability is enabled (SAI query).
+    - Parse and validate policer info.
+  - Policer-Orch
+    -  Add/Remove/Manage Policers.
+    - Validate that the policer already exists and return it's id
+    - Prevent from deleting policer that bound to ACLs
+
+##### *Image 1: Configuration Flow Overview*
+
+![alt text](Config_Flow.jpg)
+
+
+---
+### SAI API
+
+Use these **existing** SAI APIs for packet actions and policer association:
+
+| SAI Attribute                           | Description                                                                 |
+| --------------------------------------- | --------------------------------------------------------------------------- |
+| SAI_ACL_ACTION_TYPE_PACKET_ACTION       | Action type (Forward, Drop, etc) that can be taken in that ACL entry        |
+| SAI_ACL_ACTION_TYPE_SET_POLICER         | Action type (policer) that can be taken in that ACL entry                   |
+| SAI_ACL_ENTRY_ATTR_ACTION_PACKET_ACTION | Action (Forward, Drop, etc) to be executed on packets matching the ACL rule |
+| SAI_ACL_ENTRY_ATTR_ACTION_SET_POLICER   | Action (policer) to be executed on packets matching the ACL rule            |
+
+#### Usage example:
+Define the possible actions for the ACL table:
```

Comment:
these are good example, it may be better to provide examples in config_db table format